00_Base/src/services/credentials.service.ts

```diff
@@ -474,6 +477,119 @@ export class CredentialsService {
     return updatedClientInformation;
   }
 
+  async generateCredentialsTokenA(
+    credentialsRequest: CredentialsRequestDTO,
+    versionNumber: VersionNumber,
+  ): Promise<CredentialsDTO> {
+    const serverCredentialsRoleDTOs = credentialsRequest.roles;
+
+    for (const role of serverCredentialsRoleDTOs) {
+      try {
+        const storedRole =
+          await this.serverCredentialsRoleRepository.getServerCredentialsRoleByCountryCodeAndPartyId(
+            role.country_code,
+            role.party_id,
+          );
+        if (storedRole) {
+          throw new InternalServerError(
```

Comment:
why does this throw an error? the server's cpo role would be reused across many credentials, no?

Reply:
I changed the logic to update if the role exists but still validate if roles from the request body has been stored then they should belongs to the same CpoTenant. Please feel free to let me know if we should do it in another way.